### PR TITLE
Add pipeline for building PCL's environment docker image

### PR DESCRIPTION
**Pull request by kunaltyagi**
Features:
* Runs on master and PR
* Only runs when the file is modified
* Runs everyweek to keep the image updated
* No image updates on PR

*squash-worthy*

### CODE REVIEW

File: .ci/azure-pipelines/env.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    displayName: "env docker image: "
+    pool:
+      vmImage: 'ubuntu-latest'
+    strategy:
+      matrix:
+        Ubuntu 16.04:
+          UBUNTU_DISTRO: 16.04
+          tag: 16.04
+        Ubuntu 20.04:
+          UBUNTU_DISTRO: 20.04
```
Comment (taketwo) [inline .ci/azure-pipelines/env.yml:44]: Our "latest Ubuntu" image is built with `VTK_VERSION=7` argument to (somewhat) increase coverage. I'd go for 8, of course, but it's not packaged even in 20.04 :(

**Comment by kunaltyagi** [comment]
> Do we really need a daily build 

Nope. Just added to keep it up to date. Frequency doesn't matter.

File: .ci/azure-pipelines/env.yml [hunk 2]
```diff
@@ -38,24 +44,41 @@ stages:
...
     - task: Docker@2
       displayName: "Build docker image: env:$(tag)"
       inputs:
         command: build
-        arguments: --no-cache --build-arg UBUNTU_DISTRO=$(UBUNTU_DISTRO) -t $(dockerHubID)/env:$(tag)
+        arguments: |
+          --no-cache
+          --build-arg UBUNTU_DISTRO=$(UBUNTU_DISTRO)
+          --build-arg VTK_VERSION=$(VTK_VERSION)
+          --build-arg USE_CUDA=true
```
Comment (taketwo) [inline .ci/azure-pipelines/env.yml:73]:
```suggestion
          --build-arg USE_CUDA=$(USE_CUDA)
```

**Comment by kunaltyagi** [comment]
[Lots of trial and error](https://kunaltyagi.visualstudio.com/pcl/_build/results?buildId=96&view=logs&j=9206fc7c-1fc1-5f8f-29ab-987e7e61c8bc&t=de441287-6f13-51f0-88dd-fc74012310e9)

**Comment by taketwo** [comment]
Let's get a look consistent with our other pipelines:

image

0. When I setup a new pipeline on Azure, I'll call it `Docker.Env` (in line with `Build.Ubuntu`, `Build.Windows`).
1. Inside the pipeline, no need to repeat "env docker image" in the job title because it will be a part of the pipeline name.
2. No need to repeat tag in the step names, that's evident from the job name.

image

**Comment by kunaltyagi** [comment]
[How does this look?](https://kunaltyagi.visualstudio.com/pcl/_build/results?buildId=99&view=results)

**Comment by taketwo** [comment]
This looks almost perfect! It's just that "Env" is still duplicated:

image

**Comment by kunaltyagi** [comment]
So the env in the title is from title, and that can't go away. Shall I replace the other env with "Image" or "Tag" or make it empty? Are job titles allowed to be empty?

**Comment by shrijitsingh99** [comment]
> Are job titles allowed to be empty?
Shouldn't be an issue as the display name is only for visual purposes only.

We should also integrate this with stages.

**Comment by kunaltyagi** [comment]
> We should also integrate this with stages

I don't know if the images are artifacts passed across stage boundaries

**Comment by shrijitsingh99** [comment]
> We should also integrate this with stages

I meant to add it as a parallel stage, like documentation.

> I don't know if the images are artifacts passed across stage boundaries

I don't think it is passed as an artifact. Currently, to my knowledge, Azure doesn't support any official method of using a local docker image. The only way is to manage the docker container manually or some hack by pushing temporary images in the docker hub.

**Comment by kunaltyagi** [comment]
> I meant to add it as a parallel stage, like documentation

It's automatically in parallel in a diff pipeline, right?

**Comment by shrijitsingh99** [comment]
Yeah it is in a different in a pipeline but it will not show up on the stages chart:
IMG_20200405_094334

**Comment by kunaltyagi** [comment]
I'll take a look at that

**Comment by taketwo** [comment]
> Yeah it is in a different in a pipeline but it will not show up on the stages chart:

... which I think is fine, because Docker business is separate from the rest of the library. It's triggered by changes in one particular file (and also on Sundays).

**Comment by kunaltyagi** [comment]
I actually intended to test that but forgot. I'll test and report back.

**Comment by taketwo** [comment]
How is it going?

**Comment by kunaltyagi** [comment]
I don't have access to my 'build' hardware, so didn't test.

**Comment by kunaltyagi** [comment]
Only built common, geometry and a few other (non-CUDA/GPU) modules. No issues in compilation for the 16.04+CUDA-9.2

Let me add a verify stage to the CI for CI based verification

**Comment by kunaltyagi** [comment]
Do you think it makes sense to integrate all into 1 pipeline?

1. Build docker image conditionally (my addition, stage 1)
2. Use the image to compile (current pipeline), confirms docker image if it was updated
3. Push image if compile succeeded (my addition, last stage)

The only thing I don't know is if docker images are persistent over stages. 

Also, how do I run a step after all steps, like a cleanup step (especially after failure)? This is to delete the docker image if compile or push fails (to prevent bad caching on the local machine)

**Comment by taketwo** [comment]
> Do you think it makes sense to integrate all into 1 pipeline?

Maybe. Maybe not. I'd propose to start with having a separate docker pipeline. This will be an improvement already. If after some time we feel like this separation limits us, then we can consider your proposal.

**Comment by kunaltyagi** [comment]
In that case, I'll consider this ready for review. [PTAL](https://kunaltyagi.visualstudio.com/pcl/_build/results?buildId=177&view=results)

**Comment by kunaltyagi** [comment]
@taketwo it needs a [connection with the name](https://github.com/PointCloudLibrary/pcl/pull/3843/files#diff-670aa5a7c2b60449d086eb246e8adf7cR29) (and passwd) on Azure in order to run correctly

**Comment by SergioRAgostinho** [comment]
> @taketwo it needs a [connection with the name](https://github.com/PointCloudLibrary/pcl/pull/3843/files#diff-670aa5a7c2b60449d086eb246e8adf7cR29) (and passwd) on Azure in order to run correctly

Feel free to merge this at your own convenience since I'm unsure if this is ok to merge given your last comment.

File: .ci/azure-pipelines/env.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  displayName: "Sunday midnight build"
+  branches:
+    include:
+    - master
+
+resources:
+- repo: self
+
+variables:
+  dockerHub: "PersonalDockerHub"
```
Comment (kunaltyagi) [inline .ci/azure-pipelines/env.yml:29]:
```suggestion
  dockerHub: 'PointCloudLibrary@hub.docker.com'
```

**Comment by taketwo** [comment]
I'm also unsure, is there an action item for me?

**Comment by kunaltyagi** [comment]
Thanks to Sergio, who also has DockerHub access: No 😄 

The pipeline is live: https://dev.azure.com/PointCloudLibrary/pcl/_build?definitionId=11&_a=summary

**Comment by taketwo** [comment]
Awesome. Should we also setup DockerHub access for you?

**Comment by kunaltyagi** [comment]
Already done courtesy of Sergio

**Pull request merged**